File: fsharp/space-age/SpaceAge.fs
module SpaceAge

type Planet =
    | Mercury
    | Venus
    | Earth
    | Mars
    | Jupiter
    | Saturn
    | Uranus
    | Neptune

let earthSeconds = float 31557600

let computeAge (seconds: int64) (factor: float) =
    let earthYears = (float seconds) / earthSeconds
    earthYears / factor

let age (planet: Planet) (seconds: int64): float =
    match planet with
    | Mercury -> computeAge seconds 0.2408467
    | Venus -> computeAge seconds 0.61519726
    | Earth -> computeAge seconds 1.
    | Mars -> computeAge seconds 1.8808158
    | Jupiter -> computeAge seconds 11.862615
    | Saturn -> computeAge seconds 29.447498
    | Uranus -> computeAge seconds 84.016846
    | Neptune -> computeAge seconds 164.79132

File: typescript/space-age/space-age.ts
const EARTH_YEAR_SECONDS = 31_557_600

interface IPlanetMap {
  [key: string]: number
}

const PlanetMap: IPlanetMap = {
  mercury: EARTH_YEAR_SECONDS * 0.2408467,
  venus: EARTH_YEAR_SECONDS * 0.61519726,
  earth: EARTH_YEAR_SECONDS,
  mars: EARTH_YEAR_SECONDS * 1.8808158,
  jupiter: EARTH_YEAR_SECONDS * 11.862615,
  saturn: EARTH_YEAR_SECONDS * 29.447498,
  uranus: EARTH_YEAR_SECONDS * 84.016846,
  neptune: EARTH_YEAR_SECONDS * 164.79132,
};

export type Planet = keyof typeof PlanetMap;

export const age = (planet: Planet, seconds: number): number =>
  Number((seconds / PlanetMap[planet]).toFixed(2));
